content/blog/2022/headed-playwright-in-wsl-docker.md

```diff
@@ -0,0 +1,114 @@
+---
+title: Running Playwright in headed mode inside WSL Docker containers
+author: ed
+date: 2022-11-30
+tags:
+  - Article
+  - Testing
+  - Docker
+summary: |
+  Learn how to run Playwright in headed mode to interact with the browser's UI
+  from outside WSL Docker containers.
+---
+
+[Playwright](https://playwright.dev/) is a test runner that uses real browsers
+to test web applications (an alternative to tools like
+[Selenium](https://www.selenium.dev/)). By default, Playwright runs these
+browsers in headless mode, which means the pages are loaded and tested without
+opening the browser window. This is great when running entire test suites
+locally and in CI where having a bunch of rapidly opening windows would be
+pretty disconcerting. However, when it comes to writing or debugging individual
+tests, it is convenient to open the browser in *headed* mode to actually see the
+page being tested.
+
+Playwright includes the [Playwright
+Inspector](https://playwright.dev/docs/debug#playwright-inspector) to
+conveniently launch the browser in headed mode *and* a separate window to
+control test execution. The inspector is launched by setting `PWDEBUG=1` before
+calling `playwright`. The inspector is compatible with all browsers supported by
+Playwright (Safari, Chrome, and Firefox) in all major operating systems.
+Playwright also includes a `playwright open <url>` subcommand to quickly launch
+the inspector on any URL. We will use that for the rest of the article but keep
+in mind everything applies to `PWDEBUG=1` as well.
+
+A convenient [Docker image](https://playwright.dev/docs/docker) is also provided
+by Playwright and it includes all three browsers pre-installed and configured so
+you can skip the dependency installation steps. But, what happens if you try to
+run the Playwright Inspector inside a container, which normally doesn't have a
+graphical user interface?
+
+```bash
+# We expect a browser window to open and load google.com
+# Update the image tag to match your desired version of Playwright
+docker run --rm mcr.microsoft.com/playwright:v1.28.0 npx -y playwright open google.com
```

Comment:
Does this (and the other reference later) need `-focal`?

Reply:
I don't think it's important to specify the base Ubuntu version 🤷‍♂️

Reply:
Sure, definitely not as long as this command works without it 👍 (I was just comparing with the Playwright docs)